internal/server/httpapi/trigger.go

```diff
@@ -0,0 +1,398 @@
+
+			cn, ok := w.(http.CloseNotifier)
+			if !ok {
+				log.Error("failed to stream job output, could not create http.CloseNotifier")
+				http.NotFound(w, r)
+				return
+			}
+			flusher, ok := w.(http.Flusher)
+			if !ok {
+				log.Error("failed to stream job output, could not create http.Flusher")
+				http.NotFound(w, r)
+				return
+			}
+
+			// Send the initial headers saying we're gonna stream the response.
+			w.Header().Set("Transfer-Encoding", "chunked")
+			w.WriteHeader(http.StatusOK)
+			flusher.Flush()
+
+			enc := json.NewEncoder(w)
+
+			// NOTE(briancain): We skip every two jobs here because when we call RunTrigger
+			// via gRPC, it eventually queues the trigger jobs through on-demand runners, and
+			// queueJobMulti returns three jobs: StartTask, the job to be queued, and StopTask. People
+			// really only expect output from the job to be queued, so we only stream that back.
+
+			// For example, a trigger that queues 2 Waypoint operation jobs returns 6 total job ids:
+			// Job List: [ {0: StartTask, 1: WP Operation 1, 2: StopTask}, {3: StartTask, 4: WP Operation 2, 5: StopTask}, ... ]
+
+			var triggerJobs []string
+			for i := 1; i < len(jobIds); i += 3 {
+				triggerJobs = append(triggerJobs, jobIds[i])
+			}
+
+			var (
+				wg sync.WaitGroup
+				mu sync.Mutex
+			)
+
+			log.Trace("starting job stream for jobs", "total_jobs", len(triggerJobs))
+			wg.Add(len(triggerJobs))
+
+			// NOTE(briancain): This loop starts N goroutines concurrently for
+			// each trigger job to stream back to the requester.
+			// TODO(briancain): turn the http errors into Messages to send back
+			for _, jId := range triggerJobs {
+				go func(jId string) {
+					defer wg.Done()
+
+					stream, err := client.GetJobStream(ctx, &pb.GetJobStreamRequest{
+						JobId: jId,
+					})
+					if err != nil {
+						log.Error("server failed to get job stream output for trigger", "job_id", jId, "err", err)
+						http.Error(w,
+							fmt.Sprintf("server failed to obtain job stream output: %s", err),
+							http.StatusInternalServerError)
+						return
+					}
+
+					// Wait for open confirmation
+					resp, err := stream.Recv()
+					if err != nil {
+						log.Error("server failed to stream job output", "err", err)
+						http.Error(w,
+							fmt.Sprintf("server failed to receive job stream output: %s", err),
+							http.StatusInternalServerError)
+						return
+					}
+					if _, ok := resp.Event.(*pb.GetJobStreamResponse_Open_); !ok {
+						log.Error("server failed to open job stream output, got unexpected message", "event", resp.Event)
+						http.Error(w,
+							fmt.Sprintf("job stream failed to open, got unexpected message: %T", resp.Event),
+							http.StatusInternalServerError)
+						return
+					}
+
+					var (
+						jobComplete bool
+						exitCode    string
+					)
+
+					log.Trace("reading job stream for job", "job_id", jId)
+
+					// read and send the stream
+					for {
+						resp, err := stream.Recv()
+						if err != nil {
```

Comment:
We call `stream.Recv()` on line 225 above but we don't seem to do anything with the values other than error check. Is there a reason those checks can't be done in this `for` loop? If we need to get the `GetJobStreamResponse_Open_` response first, we could check for that first and then `continue` to restart the `for` loop. Would that work?

Reply:
Yep, this is because the way our job stream events work, we need to make sure the first event is always the open message. If not, we should exit right away. This is similar to how we stream job output on the CLI too. I like it separate from the `for` only because it's a required pre-condition that we only care about before reading the rest of the stream and beyond that we don't need to check for it again.

Reply:
Makes sense, thanks 👍